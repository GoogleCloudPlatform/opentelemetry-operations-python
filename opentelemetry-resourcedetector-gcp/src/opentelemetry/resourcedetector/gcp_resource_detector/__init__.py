# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
import os

import requests
from opentelemetry.context import attach, detach, set_value
from opentelemetry.sdk.resources import Resource, ResourceDetector

_GCP_METADATA_URL = (
    "http://metadata.google.internal/computeMetadata/v1/?recursive=true"
)
_GCP_METADATA_URL_HEADER = {"Metadata-Flavor": "Google"}

logger = logging.getLogger(__name__)


def _get_google_metadata_and_common_attributes():
    token = attach(set_value("suppress_instrumentation", True))
    all_metadata = requests.get(
        _GCP_METADATA_URL, headers=_GCP_METADATA_URL_HEADER
    ).json()
    detach(token)
    common_attributes = {
        "cloud.account.id": all_metadata["project"]["projectId"],
        "cloud.provider": "gcp",
        "cloud.zone": all_metadata["instance"]["zone"].split("/")[-1],
    }
    return common_attributes, all_metadata


def get_gce_resources():
    """Resource finder for common GCE attributes

    See: https://cloud.google.com/compute/docs/storing-retrieving-metadata
    """
    (
        common_attributes,
        all_metadata,
    ) = _get_google_metadata_and_common_attributes()
    common_attributes.update(
        {
            "host.id": all_metadata["instance"]["id"],
            "gcp.resource_type": "gce_instance",
        }
    )
    return common_attributes


def get_gke_resources():
    """Resource finder for GKE attributes"""

    if os.getenv("KUBERNETES_SERVICE_HOST") is None:
        return {}

    (
        common_attributes,
        all_metadata,
    ) = _get_google_metadata_and_common_attributes()

    container_name = os.getenv("CONTAINER_NAME")
    if container_name is not None:
        common_attributes["container.name"] = container_name

    # Fallback to reading namespace from a file is the env var is not set
    pod_namespace = os.getenv("NAMESPACE")
    if pod_namespace is None:
        try:
            with open(
                "/var/run/secrets/kubernetes.io/serviceaccount/namespace", "r"
            ) as namespace_file:
                pod_namespace = namespace_file.read().strip()
        except FileNotFoundError:
            pod_namespace = ""

    common_attributes.update(
        {
            "k8s.cluster.name": all_metadata["instance"]["attributes"][
                "cluster-name"
            ],
            "k8s.namespace.name": pod_namespace,
            "k8s.pod.name": os.getenv("POD_NAME", os.getenv("HOSTNAME", "")),
            "host.id": all_metadata["instance"]["id"],
            "gcp.resource_type": "gke_container",
        }
    )
    return common_attributes


def get_cloudrun_resources():
    """Resource finder for Cloud Run attributes"""

    if os.getenv("K_CONFIGURATION") is None:
        return {}

    (
        common_attributes,
        all_metadata,
    ) = _get_google_metadata_and_common_attributes()

    faas_name = os.getenv("K_SERVICE")
    if faas_name is not None:
        common_attributes["faas.name"] = str(faas_name)

    faas_version = os.getenv("K_REVISION")
    if faas_version is not None:
        common_attributes["faas.version"] = str(faas_version)

    common_attributes.update(
        {
            "cloud.platform": "gcp_cloud_run",
            "cloud.region": all_metadata["instance"]["region"].split("/")[-1],
            "faas.instance": all_metadata["instance"]["id"],
            "gcp.resource_type": "cloud_run",
        }
    )
    return common_attributes


def get_cloudfunctions_resources():
    """Resource finder for Cloud Functions attributes"""

    if os.getenv("FUNCTION_TARGET") is None:
        return {}

    (
        common_attributes,
        all_metadata,
    ) = _get_google_metadata_and_common_attributes()

    faas_name = os.getenv("K_SERVICE")
    if faas_name is not None:
        common_attributes["faas.name"] = str(faas_name)

    faas_version = os.getenv("K_REVISION")
    if faas_version is not None:
        common_attributes["faas.version"] = str(faas_version)

    common_attributes.update(
        {
            "cloud.platform": "gcp_cloud_functions",
            "cloud.region": all_metadata["instance"]["region"].split("/")[-1],
            "faas.instance": all_metadata["instance"]["id"],
            "gcp.resource_type": "cloud_functions",
        }
    )
    return common_attributes


# Order here matters. Since a GKE_CONTAINER is a specialized type of GCE_INSTANCE
# We need to first check if it matches the criteria for being a GKE_CONTAINER
# before falling back and checking if its a GCE_INSTANCE.
# This list should be sorted from most specialized to least specialized.
_RESOURCE_FINDERS = [
    ("gke_container", get_gke_resources),
    ("cloud_run", get_cloudrun_resources),
    ("cloud_functions", get_cloudfunctions_resources),
    ("gce_instance", get_gce_resources),
]


class NoGoogleResourcesFound(Exception):
    pass


class GoogleCloudResourceDetector(ResourceDetector):
    def __init__(self, raise_on_error=False):
        super().__init__(raise_on_error)
        self.cached = False
        self.gcp_resources = {}

    def detect(self) -> "Resource":
        if not self.cached:
            self.cached = True
            for resource_type, resource_finder in _RESOURCE_FINDERS:
                try:
                    found_resources = resource_finder()
                # pylint: disable=broad-except
                except Exception as ex:
                    logger.warning(
                        "Exception %s occured attempting %s resource detection",
                        ex,
                        resource_type,
                    )
                    found_resources = None
                if found_resources:
                    self.gcp_resources = found_resources
                    break
        if not self.gcp_resources:
            raise NoGoogleResourcesFound()
        return Resource(self.gcp_resources)
