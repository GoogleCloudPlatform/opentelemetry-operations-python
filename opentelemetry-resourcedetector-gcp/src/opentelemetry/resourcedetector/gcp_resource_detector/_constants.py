# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# TODO: use opentelemetry-semantic-conventions package for these constants once it has
# stabilized. Right now, pinning an unstable version would cause dependency conflicts for
# users so these are copied in.
class ResourceAttributes:
    AWS_EC2 = "aws_ec2"
    CLOUD_ACCOUNT_ID = "cloud.account.id"
    CLOUD_AVAILABILITY_ZONE = "cloud.availability_zone"
    CLOUD_PLATFORM_KEY = "cloud.platform"
    CLOUD_PROVIDER = "cloud.provider"
    CLOUD_REGION = "cloud.region"
    GCP_COMPUTE_ENGINE = "gcp_compute_engine"
    GCP_KUBERNETES_ENGINE = "gcp_kubernetes_engine"
    HOST_ID = "host.id"
    HOST_NAME = "host.name"
    HOST_TYPE = "host.type"
    K8S_CLUSTER_NAME = "k8s.cluster.name"
    K8S_CONTAINER_NAME = "k8s.container.name"
    K8S_NAMESPACE_NAME = "k8s.namespace.name"
    K8S_NODE_NAME = "k8s.node.name"
    K8S_POD_NAME = "k8s.pod.name"
    SERVICE_INSTANCE_ID = "service.instance.id"
    SERVICE_NAME = "service.name"
    SERVICE_NAMESPACE = "service.namespace"
    FAAS_INSTANCE = "faas.instance"
    FAAS_NAME = "faas.name"


AWS_ACCOUNT = "aws_account"
AWS_EC2_INSTANCE = "aws_ec2_instance"
CLUSTER_NAME = "cluster_name"
CONTAINER_NAME = "container_name"
GCE_INSTANCE = "gce_instance"
GENERIC_NODE = "generic_node"
GENERIC_TASK = "generic_task"
INSTANCE_ID = "instance_id"
JOB = "job"
K8S_CLUSTER = "k8s_cluster"
K8S_CONTAINER = "k8s_container"
K8S_NODE = "k8s_node"
K8S_POD = "k8s_pod"
LOCATION = "location"
NAMESPACE = "namespace"
NAMESPACE_NAME = "namespace_name"
NODE_ID = "node_id"
NODE_NAME = "node_name"
POD_NAME = "pod_name"
REGION = "region"
TASK_ID = "task_id"
ZONE = "zone"
UNKNOWN_SERVICE_PREFIX = "unknown_service"
