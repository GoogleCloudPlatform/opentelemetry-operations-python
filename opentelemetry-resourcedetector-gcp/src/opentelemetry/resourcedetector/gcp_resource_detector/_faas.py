# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Implementation in this file copied from
# https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/v1.8.0/detectors/gcp/faas.go

import os

from opentelemetry.resourcedetector.gcp_resource_detector import _metadata

_CLOUD_RUN_CONFIG_ENV = "K_CONFIGURATION"
_CLOUD_FUNCTION_TARGET_ENV = "FUNCTION_TARGET"
_FAAS_SERVICE_ENV = "K_SERVICE"
_FAAS_REVISION_ENV = "K_REVISION"


def on_cloud_run() -> bool:
    return _CLOUD_RUN_CONFIG_ENV in os.environ


def on_cloud_functions() -> bool:
    return _CLOUD_FUNCTION_TARGET_ENV in os.environ


def faas_name() -> str:
    """The name of the Cloud Run or Cloud Function.

    Check that on_cloud_run() or on_cloud_functions() is true before calling this, or it may
    throw exceptions.
    """
    return os.environ[_FAAS_SERVICE_ENV]


def faas_version() -> str:
    """The version/revision of the Cloud Run or Cloud Function.

    Check that on_cloud_run() or on_cloud_functions() is true before calling this, or it may
    throw exceptions.
    """
    return os.environ[_FAAS_REVISION_ENV]


def faas_instance() -> str:
    return str(_metadata.get_metadata()["instance"]["id"])


def faas_cloud_region() -> str:
    region = _metadata.get_metadata()["instance"]["region"]
    return region[region.rfind("/") + 1 :]
