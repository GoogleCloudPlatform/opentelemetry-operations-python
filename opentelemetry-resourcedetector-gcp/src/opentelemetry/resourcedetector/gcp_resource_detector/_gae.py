# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Implementation in this file copied from
# https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/v1.8.0/detectors/gcp/app_engine.go

import os

from opentelemetry.resourcedetector.gcp_resource_detector import (
    _faas,
    _gce,
    _metadata,
)

_GAE_SERVICE_ENV = "GAE_SERVICE"
_GAE_VERSION_ENV = "GAE_VERSION"
_GAE_INSTANCE_ENV = "GAE_INSTANCE"
_GAE_ENV = "GAE_ENV"
_GAE_STANDARD = "standard"


def on_app_engine_standard() -> bool:
    return os.environ.get(_GAE_ENV) == _GAE_STANDARD


def on_app_engine() -> bool:
    return _GAE_SERVICE_ENV in os.environ


def service_name() -> str:
    """The service name of the app engine service.

    Check that ``on_app_engine()`` is true before calling this, or it may throw exceptions.
    """
    return os.environ[_GAE_SERVICE_ENV]


def service_version() -> str:
    """The service version of the app engine service.

    Check that ``on_app_engine()`` is true before calling this, or it may throw exceptions.
    """
    return os.environ[_GAE_VERSION_ENV]


def service_instance() -> str:
    """The service instance of the app engine service.

    Check that ``on_app_engine()`` is true before calling this, or it may throw exceptions.
    """
    return os.environ[_GAE_INSTANCE_ENV]


def flex_availability_zone_and_region() -> _gce.ZoneAndRegion:
    """The zone and region in which this program is running.

    Check that ``on_app_engine()`` is true before calling this, or it may throw exceptions.
    """
    return _gce.availability_zone_and_region()


def standard_availability_zone() -> str:
    """The zone the app engine service is running in.

    Check that ``on_app_engine_standard()`` is true before calling this, or it may throw exceptions.
    """
    zone = _metadata.get_metadata()["instance"]["zone"]
    # zone is of the form "projects/233510669999/zones/us15"
    return zone[zone.rfind("/") + 1 :]


def standard_cloud_region() -> str:
    """The region the app engine service is running in.

    Check that ``on_app_engine_standard()`` is true before calling this, or it may throw exceptions.
    """
    return _faas.faas_cloud_region()
