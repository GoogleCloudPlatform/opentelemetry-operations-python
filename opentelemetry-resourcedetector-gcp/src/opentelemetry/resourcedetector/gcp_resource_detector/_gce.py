# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
import re
from dataclasses import dataclass

from opentelemetry.resourcedetector.gcp_resource_detector import _metadata

# Format described in
# https://cloud.google.com/compute/docs/metadata/default-metadata-values#vm_instance_metadata
_ZONE_REGION_RE = re.compile(
    r"projects\/\d+\/zones\/(?P<zone>(?P<region>\w+-\w+)-\w+)"
)

_logger = logging.getLogger(__name__)


def on_gce() -> bool:
    try:
        _metadata.get_metadata()["instance"]["machineType"]
    except (_metadata.MetadataAccessException, KeyError):
        _logger.debug(
            "Could not fetch metadata attribute instance/machineType, "
            "assuming not on GCE.",
            exc_info=True,
        )
        return False
    return True


def host_type() -> str:
    return _metadata.get_metadata()["instance"]["machineType"]


def host_id() -> str:
    return _metadata.get_metadata()["instance"]["id"]


def host_name() -> str:
    return _metadata.get_metadata()["instance"]["name"]


@dataclass
class ZoneAndRegion:
    zone: str
    region: str


def availability_zone_and_region() -> ZoneAndRegion:
    full_zone = _metadata.get_metadata()["instance"]["zone"]
    match = _ZONE_REGION_RE.search(full_zone)
    if not match:
        raise Exception(
            "zone was not in the expected format: "
            f"projects/PROJECT_NUM/zones/COUNTRY-REGION-ZONE. Got {full_zone}"
        )

    return ZoneAndRegion(
        zone=match.group("zone"), region=match.group("region")
    )
