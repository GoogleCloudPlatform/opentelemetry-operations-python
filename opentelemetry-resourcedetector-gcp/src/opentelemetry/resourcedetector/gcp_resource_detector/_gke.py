# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
from dataclasses import dataclass

from opentelemetry.resourcedetector.gcp_resource_detector import (
    _gce,
    _metadata,
)

# TODO: remove when Python 3.7 is dropped
from typing_extensions import Literal

KUBERNETES_SERVICE_HOST_ENV = "KUBERNETES_SERVICE_HOST"


def on_gke() -> bool:
    return os.environ.get(KUBERNETES_SERVICE_HOST_ENV) is not None


def host_id() -> str:
    return _gce.host_id()


def cluster_name() -> str:
    return _metadata.get_metadata()["instance"]["attributes"]["cluster-name"]


@dataclass
class ZoneOrRegion:
    type: Literal["zone", "region"]
    value: str


def availability_zone_or_region() -> ZoneOrRegion:
    cluster_location = _metadata.get_metadata()["instance"]["attributes"][
        "cluster-location"
    ]
    hyphen_count = cluster_location.count("-")
    if hyphen_count == 1:
        return ZoneOrRegion(type="region", value=cluster_location)
    if hyphen_count == 2:
        return ZoneOrRegion(type="zone", value=cluster_location)
    raise Exception(
        f"unrecognized format for cluster location: {cluster_location}"
    )
