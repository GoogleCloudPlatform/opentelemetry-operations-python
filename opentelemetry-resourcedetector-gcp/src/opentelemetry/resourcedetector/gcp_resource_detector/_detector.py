# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Mapping

from opentelemetry.resourcedetector.gcp_resource_detector import (
    _gce,
    _gke,
    _metadata,
)
from opentelemetry.resourcedetector.gcp_resource_detector._constants import (
    ResourceAttributes,
)
from opentelemetry.sdk.resources import Resource, ResourceDetector
from opentelemetry.util.types import AttributeValue


class GoogleCloudResourceDetector(ResourceDetector):
    def detect(self) -> Resource:
        if not _metadata.is_available():
            return Resource.get_empty()

        if _gke.on_gke():
            return _gke_resource()
        if _gce.on_gce():
            return _gce_resource()

        return Resource.get_empty()


def _gke_resource() -> Resource:
    zone_or_region = _gke.availability_zone_or_region()
    zone_or_region_key = (
        ResourceAttributes.CLOUD_AVAILABILITY_ZONE
        if zone_or_region.type == "zone"
        else ResourceAttributes.CLOUD_REGION
    )
    return _make_resource(
        {
            ResourceAttributes.CLOUD_PLATFORM_KEY: ResourceAttributes.GCP_KUBERNETES_ENGINE,
            zone_or_region_key: zone_or_region.value,
            ResourceAttributes.K8S_CLUSTER_NAME: _gke.cluster_name(),
            ResourceAttributes.HOST_ID: _gke.host_id(),
        }
    )


def _gce_resource() -> Resource:
    zone_and_region = _gce.availability_zone_and_region()
    return _make_resource(
        {
            ResourceAttributes.CLOUD_PLATFORM_KEY: ResourceAttributes.GCP_COMPUTE_ENGINE,
            ResourceAttributes.CLOUD_AVAILABILITY_ZONE: zone_and_region.zone,
            ResourceAttributes.CLOUD_REGION: zone_and_region.region,
            ResourceAttributes.HOST_TYPE: _gce.host_type(),
            ResourceAttributes.HOST_ID: _gce.host_id(),
            ResourceAttributes.HOST_NAME: _gce.host_name(),
        }
    )


def _make_resource(attrs: Mapping[str, AttributeValue]) -> Resource:
    return Resource.create(
        {
            ResourceAttributes.CLOUD_PROVIDER: "gcp",
            ResourceAttributes.CLOUD_ACCOUNT_ID: _metadata.get_metadata()[
                "project"
            ]["projectId"],
            **attrs,
        }
    )
