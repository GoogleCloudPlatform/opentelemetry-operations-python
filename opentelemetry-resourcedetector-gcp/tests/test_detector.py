# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
from unittest.mock import Mock

import pytest
import requests
from opentelemetry.resourcedetector.gcp_resource_detector import _metadata
from opentelemetry.resourcedetector.gcp_resource_detector._detector import (
    GoogleCloudResourceDetector,
)


@pytest.fixture(name="reset_cache")
def fixture_reset_cache():
    yield
    _metadata.get_metadata.cache_clear()
    _metadata.is_available.cache_clear()


@pytest.fixture(name="fake_get")
def fixture_fake_get(monkeypatch: pytest.MonkeyPatch):
    mock = Mock()
    monkeypatch.setattr(requests, "get", mock)
    return mock


@pytest.fixture(name="fake_metadata")
def fixture_fake_metadata(fake_get: Mock):
    json = {"instance": {}, "project": {}}
    fake_get().json.return_value = json
    return json


# Reset stuff before every test
# pylint: disable=unused-argument
@pytest.fixture(autouse=True)
def autouse(reset_cache, fake_get, fake_metadata, fake_environ):
    pass


def test_detects_empty_when_not_available(snapshot, fake_get: Mock):
    fake_get.side_effect = requests.HTTPError()
    assert dict(GoogleCloudResourceDetector().detect().attributes) == snapshot


def test_detects_empty_as_fallback(snapshot):
    assert dict(GoogleCloudResourceDetector().detect().attributes) == snapshot


def test_detects_gce(snapshot, fake_metadata: _metadata.Metadata):
    fake_metadata.update(
        {
            "project": {"projectId": "fakeProject"},
            "instance": {
                "name": "fakeName",
                "id": "fakeId",
                "machineType": "fakeMachineType",
                "zone": "projects/233510669999/zones/us-east4-b",
            },
        }
    )

    assert dict(GoogleCloudResourceDetector().detect().attributes) == snapshot


@pytest.mark.parametrize(
    "cluster_location",
    (
        pytest.param("us-east4", id="regional"),
        pytest.param("us-east4-b", id="zonal"),
    ),
)
def test_detects_gke(
    cluster_location: str, snapshot, fake_metadata: _metadata.Metadata
):
    os.environ["KUBERNETES_SERVICE_HOST"] = "fakehost"
    fake_metadata.update(
        # All the same attributes as GCE
        {
            "project": {"projectId": "fakeProject"},
            "instance": {
                "name": "fakeName",
                "id": "fakeId",
                "machineType": "fakeMachineType",
                "zone": "projects/233510669999/zones/us-east4-b",
                # Plus some attributes
                "attributes": {
                    "cluster-name": "fakeClusterName",
                    "cluster-location": cluster_location,
                },
            },
        }
    )

    assert dict(GoogleCloudResourceDetector().detect().attributes) == snapshot
