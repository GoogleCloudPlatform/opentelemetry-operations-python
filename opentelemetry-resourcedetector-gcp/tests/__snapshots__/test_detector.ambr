# name: test_detects_empty_as_fallback
  dict({
  })
# ---
# name: test_detects_empty_when_not_available
  dict({
  })
# ---
# name: test_detects_gce
  dict({
    'cloud.account.id': 'fakeProject',
    'cloud.availability_zone': 'us-east4-b',
    'cloud.platform': 'gcp_compute_engine',
    'cloud.provider': 'gcp',
    'cloud.region': 'us-east4',
    'host.id': 'fakeId',
    'host.name': 'fakeName',
    'host.type': 'fakeMachineType',
    'service.name': 'unknown_service',
    'telemetry.sdk.language': 'python',
    'telemetry.sdk.name': 'opentelemetry',
    'telemetry.sdk.version': '1.20.0',
  })
# ---
# name: test_detects_gke[regional]
  dict({
    'cloud.account.id': 'fakeProject',
    'cloud.platform': 'gcp_kubernetes_engine',
    'cloud.provider': 'gcp',
    'cloud.region': 'us-east4',
    'host.id': 'fakeId',
    'k8s.cluster.name': 'fakeClusterName',
    'service.name': 'unknown_service',
    'telemetry.sdk.language': 'python',
    'telemetry.sdk.name': 'opentelemetry',
    'telemetry.sdk.version': '1.20.0',
  })
# ---
# name: test_detects_gke[zonal]
  dict({
    'cloud.account.id': 'fakeProject',
    'cloud.availability_zone': 'us-east4-b',
    'cloud.platform': 'gcp_kubernetes_engine',
    'cloud.provider': 'gcp',
    'host.id': 'fakeId',
    'k8s.cluster.name': 'fakeClusterName',
    'service.name': 'unknown_service',
    'telemetry.sdk.language': 'python',
    'telemetry.sdk.name': 'opentelemetry',
    'telemetry.sdk.version': '1.20.0',
  })
# ---
